extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,258 @@
+                      }
+                      rule.encoded_credentials.insert(
+                          Base64::encode(credential.first.value().data(),
+                                         credential.first.value().size()));
+                      return true;
+                    })) {
+              LOG_WARN("Failed to parse configuration for credentials.");
+              return false;
+            }
+
+            if (match == "prefix") {
+              rule.pattern = Prefix;
+            } else if (match == "exact") {
+              rule.pattern = Exact;
+            } else if (match == "suffix") {
+              rule.pattern = Suffix;
+            }
+            rule.request_path = request_path;
+            for (auto& method : request_methods) {
+              basic_auth_configuration_[method].push_back(rule);
+            }
+            return true;
+          })) {
+    LOG_WARN(absl::StrCat("cannot parse plugin configuration JSON string: ",
+                          configuration_data->view()));
+  }
+  return true;
+}
+
+FilterHeadersStatus PluginRootContext::credentialsCheck(
+    const BasicAuthConfigRule& rule, std::string_view authorization_header) {
+  // Check if the Basic auth header starts with "Basic "
+  if (!absl::ConsumePrefix(&authorization_header, "Basic ")) {
+    deniedNoBasicAuthData();
+    return FilterHeadersStatus::StopIteration;
+  }
+  auto auth_credential_iter =
+      rule.encoded_credentials.find(std::string(authorization_header));
+  // Check if encoded credential is part of the encoded_credentials
+  // set from our container to grant or deny access.
+  if (auth_credential_iter == rule.encoded_credentials.end()) {
+    deniedInvalidCredentials();
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus PluginContext::onRequestHeaders(uint32_t, bool) {
+  auto basic_auth_configuration = rootContext()->basicAuthConfigurationValue();
+  auto request_path_header = getRequestHeader(":path");
+  std::string_view request_path = request_path_header->view();
+  auto method = getRequestHeader(":method")->toString();
+  auto method_iter = basic_auth_configuration.find(method);
+  // First we check if the request method is present in our container
+  if (method_iter != basic_auth_configuration.end()) {
+    // We iterate through our vector of struct in order to find if the
+    // request_path according to given match patterns, is part of the plugin's
+    // configuration data. If that's the case we check the credentials
+    FilterHeadersStatus header_status = FilterHeadersStatus::Continue;
+    for (auto& rules : basic_auth_configuration[method]) {
+      if (rules.pattern == PluginRootContext::MATCH_TYPE::Prefix) {
+        if (absl::StartsWith(request_path, rules.request_path)) {
+          auto authorization_header = getRequestHeader("authorization");
+          std::string_view authorization = authorization_header->view();
```

Comment:
This will fetch authorization header multiple times and I guess you are doing it because consumePrefix/Suffix will change the string?

Absl has a function called stripPrefix/Suffix and I think it will leave the string intact, so that you can pass in authorization string_view without fetching it repeatedly.

Reply:
okok sure!